_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+better or for worse).
+
+In bounds checking, the compiler can see the values involved in the comparisons, as they come from
+constants or other values related to the array. For instance when iterating through a span, stopping
+before the span's `length`, the compiler can see that every index into the span is less than its
+`length` fairly trivially. This usually means the compiler removes the bounds checks. So no runtime
+overhead, yay!
+
+But there's another type of overhead that comes not from the bounds checks in view types like
+`std::span` or `std::string_view`, but from the change to using those types at all.
+
+Here's a commit that converts a bunch of functions to use `base::StringPiece`, the Chromium string
+view type which is much like `std::string_view`:
+[https://chromium-review.googlesource.com/c/chromium/src/+/4575826](
+https://chromium-review.googlesource.com/c/chromium/src/+/4575826).
+
+The functions were previously overloaded to take `const std::string&` or `const char*`. The
+`const char*` overload presumes that:
+- the string being passed in is null terminated, and
+- the function should be consuming the full string up to the null character.
+
+These assumptions are easily wrong when working with view types that encode a bounded view of an
+array. Calling these functions requires converting that bounded string view into a `std::string`
+which involves generating a constructor, and quite possibly a heap allocation. But
+many callers instead convert their bounded string view into a `const char*` through `.data()`
+instead, which is a bug. Props to [@davidben](https://github.com/davidben) for noticing this common
+anti-pattern in Chromium and working to fix them.
+
+So it appears at first glance to be a win to change these functions to take a string view type, in
+this case `base::StringPiece`. For callers with a `std::string`, it will copy the pointer and length
+into the `base::StringPiece`. For callers with a bounded string view type, they don't need to throw
+away the bounds or heap allocate. Yet this created a pretty large binary size regression for doing
+almost nothing, and trying to eliminate heap allocations.
+
+Chromium is an interesting test bed because it is an _enormous_ codebase. So when you make a small
+change, it can easily be amplified immensely. In this case, we discovered that all those calls to a
+`base::StringPiece` constructor grow the binary by 36KB. For unknown reasons, the
+`base::StringPiece` constructor gets inlined, while the `std::string_view` constructor does not, so
+you get less binary size growth with that type (and we'll look at marking `base::StringPiece` with a
+`noinline` attribute).
+
+But in our experience, we've seen that binary size correlates quite well with performance as well.
+So *this also implies a performance degredation by moving from pointers to a bounded view type*. And
+herein lies the problem. Even if a codebase successfully flip libc++ hardening on for their

```

Comment:
typo: flips

This sentence could use some unpacking. Maybe something like: "If developers turn on libc++ hardening in their codebase, there may be a different kind of performance regression even though the compiler usually elides the bounds checks."

Reply:
It's slightly different in that the hardening skates by thanks to elision. But rolling it out further by converting to span is the more costly thing.

Reply:
In our experience, binary size growth correlates well with performance loss, and in this case we
see strictly more codegen which is certainly going to be visible at scale in performance testing.
So *this also implies a performance degradation by moving from pointers to a bounded view type*.
And herein lies the problem. A project may successfully flip libc++ hardening on for their
`std::span` types without seeing much of a performance impact, because the majority of bounds
checks are elided. But then as they convert more pointers to `std::span` to increase their
bounds safety coverage, the use of `std::span` will introduce a different kind of performance
(and binary size) regression.

Reply:
Further rewording:

> In Chromium, when a change impacts performance, we can often just look for where binary size grew
> to track down where the regression lies. So we've seen binary size growth correlate well with
> performance loss. In this case we see strictly more codegen at each call site, which can certainly
> become visible at scale in performance testing.